src/clojupyter/kernel/comm_atom.clj

```diff
@@ -25,131 +25,116 @@
-    "Remove watch added with key `key` from `comm-atom`.  Returns `comm-atom`."))
-
-(declare agentfld comm-atom? send-comm-msg! send-comm-open! simple-fmt update-agent!)
+  (sync-state [comm-atom]
+    "A map of attributes to be sent to front-end")
+  (close! [comm-atom]
+    "Removes the comm-atom from the global state and sends COMM-CLOSE to the front end."))
+
+(declare comm-atom? send-comm-state! send-comm-open! simple-fmt jsonable?)
 
 (deftype CommAtom
-    [comm-state_ jup_ target-name_ reqmsg_ cid_ agent_]
+    [comm-id target sync-keys comm-state_]
 
   comm-atom-proto
-  (target [_]
-    target-name_)
-  (comm-id [_]
-    cid_)
-  (model-ref [_]
-    (str "IPY_MODEL_" cid_))
-  (origin-message [_]
-    reqmsg_)
-  (state-set! [comm-atom comm-state]
-    (assert (map? comm-state))
-    (reset! comm-state_ comm-state)
-    (update-agent! comm-atom comm-state)
-    (send-comm-msg! comm-atom comm-state)
-    comm-atom)
-  (state-update! [comm-atom comm-state]
-    (assert (map? comm-state))
-    (state-set! comm-atom (merge @comm-atom comm-state)))
-  (watch [comm-atom key f]
-    (assert (fn? f))
-    (add-watch (agentfld comm-atom) key f))
-  (unwatch [comm-atom key]
-    (remove-watch (agentfld comm-atom) key))
-
+  (sync-state [_]
+    (select-keys @comm-state_ sync-keys))
+  (close! [comm-atom]
+    (let [content (msgs/comm-close-content comm-id {})
+          jup (:jup @state/STATE)
+          {origin-message :req-message :or {origin-message {:header {}}}} (first (:cur-ctx @state/STATE))]
+      (jup/send!! jup :iopub_port origin-message msgs/COMM-CLOSE MESSAGE-METADATA content)
+      (swap! state/STATE update :comms dissoc comm-id)
+      nil))
+
+  java.io.Closeable
+  (close [comm-atom]
+    (msgs/leaf-paths comm-atom? #(.close %) (sync-state comm-atom))
+    (close! comm-atom))
 
   mc/PMimeConvertible
   (to-mime [_]
     (u/json-str {:text/plain
-                 ,, (str "[" cid_ "]=" @agent_)
+                 ,, (str "[" comm-id "]=" (:_model_name @comm-state_))
                  :application/vnd.jupyter.widget-view+json
                  ,, {:version_major WIDGET-VERSION-MAJOR
                      :version_minor WIDGET-VERSION-MINOR
-                     :model_id cid_}}))
+                     :model_id comm-id}}))
 
   (toString [comm-atom]
     (simple-fmt comm-atom))
 
   clojure.lang.IAtom
   (compareAndSet [comm-atom old new]
-    (if (compare-and-set! comm-state_ old new)
-      (do (state-set! comm-atom @comm-state_)
+    (assert (jsonable? (select-keys new sync-keys)))
+    (if (= @comm-atom old)
+      (do (reset! comm-atom new)
           true)
       false))
   (reset [comm-atom v]
-    (state-set! comm-atom v)
+    (assert (jsonable? (select-keys v sync-keys)))
+    (send-comm-state! comm-atom (select-keys v sync-keys))
+    (reset! comm-state_ v)
     v)
   (swap [comm-atom f]
-    (state-set! comm-atom (.swap ^Atom comm-state_ f)))
+    (let [state @comm-atom
+          n-state (f state)]
```

Comment:
I think we might be better off to swap the comm-state_ atom at this point to get the new-state instead of using `compare-and-set!` .
``` clojure
n-state (swap! comm-state_ f)]
```

Also, since at this point we have the old state and the new state, we can compute the deltas and only send those on the wire. For now, on every swap of CommsAtom we send the whole widget state and that can be very noisy, since the state maps can get arbitrarily wide and deep.

We can also do the same thing for reset, getting the initial state and only sending the deltas in a `msg-update`.

Reply:
> I think we might be better off to swap the comm-state_ atom at this point to get the new-state instead of using compare-and-set! .

Agreed. My read of this code is that `compare-and-set!` is not being used correctly. To work correctly in a concurrent system it should be called in a loop until it succeeds, like: https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Atom.java#L33 

If you use `swap!` instead this is taken care of so 👍🏻

Reply:
Fixed in 0f0ea21